rfcs/20241205-graph-api-causal-mask/README.md

```diff
@@ -0,0 +1,485 @@
+| :---- | :------------ | :------------------- |
+| 0     | `dst`         | Required             |
+
+##### Supported data types
+
+GreaterThan operation supports the following data type combinations.
+
+| src  | dst  |
+| :--- | :--- |
+| f32  | f32  |
+| f16  | f16  |
+| bf16 | bf16 |
+
+#### Constant Op Support
+
+`Constant` operation takes a scalar as an attribute, and converts it to a tensor,
+which will then be passed to *Add* and *Sub* operations in the case of bottom-right
+causal mask to provide the information of s_q and s_kv. These values are needed during
+compilation stage to generate optimized kernels.
+
+| Attribute Name | Description | Value Type | Supported Values | Required or Optional |
+| :------------- | :---------- | :--------- | :--------------- | :------------------- |
+| [scalar] (@ref dnnl::graph::op::attr::scalar) | Specifies the constant value. | f32, bf16, f16 | arbitrary valid floating point value | Required             |
+
+##### Execution arguments
+
+##### Inputs
+
+N/A
+
+##### Outputs
+
+| Index | Argument Name | Required or Optional |
+| :---- | :------------ | :------------------- |
+| 0     | `dst`         | Required             |
+
+##### Supported data types
+
+`Constant` operation supports the following data type combinations.
+
+| Attribute | dst  |
+| :-------- | :--- |
+| f32       | f32  |
+| f16       | f16  |
+| bf16      | bf16 |
+
+#### Summary of Option 1
+
+Pros:
+
+- The newly added operations are general-purpose and can be used in other
+scenarios beyond causal masking.
+- The subgraph approach provides flexibility to extend to other masking patterns
+by composing basic operations.
+- This approach aligns with cuDNN backend API design, promoting consistency across
+different libraries and frameworks.
+
+Cons:
+
+- Adding multiple operations for masking increases the overall graph complexity.
+- The library needs to add three new operations to support a single causal mask,
+which may be excessive.
+- This approach may require additional logical operations (e.g., LessThan, And)
+to be added if other mask options are needed in the future.
+- More user code is required to construct the causal mask subgraph, which adds
+complexity on the user side. This could potentially increase the learning curve and
+the likelihood of errors in implementation.
+
+#### Option 2: Support implicit causal mask through a single op
+
+Another option is to introduce a dedicated operation for mask usage.
+
+![causal mask op](images/causal_mask_op.png)
+
+#### Mask Op Support
+
+`Mask` operation masks the input tensor based on attribute settings.
+
+
+| Attribute Name | Description | Value Type | Supported Values | Required or Optional |
+| :------------- | :---------- | :--------- | :--------------- | :------------------- |
+| [mask_option] (@ref dnnl::graph::op::attr::mask_option) | What kind of mask to apply | string     | causal_mask (default), causal_mask_bottom_right | Required             |
```

Comment:
nit: it seems `mask_type` can be a better attribute name.

I assume that the purpose is to extend the value of this attribute to support more mask types, eg. padding mask, etc. Can you explain how this op will be defined for them? Do them require more inputs/outputs rather than a single attribute value change?

Reply:
BTW, have you considered to let the new op just "generate mask" and keep the Select op which is already supported?

Reply:
For more mask types, this op needs to be extended, e.g.:
1. for padding mask, besides adding a new `mask_type`, actual sequence length of KV and Q are needed.
2. for sliding window mask, an attribute of window size is needed.
3. for alibi mask, an attribute of alibi slope is needed.

Regarding the new op of "generate mask" (consider it to be named as GenMask), compared with the Mask op, GenMask needs an input of shape, the value of the input can only be known at execution time, while in the Mask op, the value of the shape can be got from the logical tensor of `src`, which is at compilation stage. We want to have the shape information earlier for some potential optimization opportunities, so Mask op is proposed here.